crates/model/src/instruments/mod.rs

```diff
@@ -151,31 +278,88 @@ pub trait Instrument: 'static + Send {
         price: Price,
         use_quote_for_inverse: Option<bool>,
     ) -> Money {
-        let use_quote_for_inverse = use_quote_for_inverse.unwrap_or(false);
-        let (amount, currency) = if self.is_inverse() {
-            if use_quote_for_inverse {
-                (quantity.as_f64(), self.quote_currency())
+        let use_quote = use_quote_for_inverse.unwrap_or(false);
+        if self.is_inverse() {
+            if use_quote {
+                Money::new(quantity.as_f64(), self.quote_currency())
             } else {
-                let amount =
-                    quantity.as_f64() * self.multiplier().as_f64() * (1.0 / price.as_f64());
-                let currency = self
+                let amt = quantity.as_f64() * self.multiplier().as_f64() * (1.0 / price.as_f64());
+                let ccy = self
                     .base_currency()
-                    .expect("Error: no base currency for notional calculation");
-                (amount, currency)
+                    .expect("inverse instrument without base currency");
+                Money::new(amt, ccy)
             }
         } else {
-            let amount = quantity.as_f64() * self.multiplier().as_f64() * price.as_f64();
-            let currency = self.quote_currency();
-            (amount, currency)
-        };
-
-        Money::new(amount, currency)
+            let amt = quantity.as_f64() * self.multiplier().as_f64() * price.as_f64();
+            Money::new(amt, self.quote_currency())
+        }
     }
 
     /// Returns the equivalent quantity of the base asset.
     fn calculate_base_quantity(&self, quantity: Quantity, last_px: Price) -> Quantity {
-        let value = quantity.as_f64() * (1.0 / last_px.as_f64());
-        Quantity::new(value, self.size_precision())
+        Quantity::new(quantity.as_f64() / last_px.as_f64(), self.size_precision())
+    }
+
+    fn next_bid_price(&self, value: f64, n: i32) -> Option<Price> {
+        let px = if let Some(scheme) = self.tick_scheme() {
+            scheme.next_bid_price(value, n, self.price_precision())?
+        } else {
+            let inc = self.price_increment().as_f64();
+            if inc <= 0.0 {
```

Comment:
Prices can potentially be negative (options, spreads, commodities etc).

Reply:
adjusted, thx